### PR TITLE
CPLP-1076 Get document endpoints

### PR DESCRIPTION
**Pull request by Phil91**
Depending on #93

### CODE REVIEW

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/PortalRepositories.cs [hunk 1]
```diff
@@ -44,6 +44,10 @@ public RepositoryType GetInstance<RepositoryType>()
         {
             return To<RepositoryType>(new UserRolesRepository(_dbContext));
         }
+        else if (repositoryType == typeof(IDocumentRepository))
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/PortalRepositories.cs:47]: Please move this to line 31, then the types are in alphabetical order which makes it easier to see whether a particular type has already been included.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/DocumentsBusinessLogic.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    /// <inheritdoc />
+    public async Task<(string fileName, byte[] content)> GetDocumentAsync(Guid documentId, string iamUserId)
+    {
+        var document = await _documentRepository.GetDocumentByIdAsync(documentId).ConfigureAwait(false);
+        if (document is null)
+        {
+            throw new NotFoundException("No document with the given id was found.");
+        }
+
+        return (document.Documentname, Convert.FromBase64String(document.DocumentContent));
```
Comment (ntruchsess) [inline src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/DocumentsBusinessLogic.cs:31]: document-content is byte[] not base64

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/DocumentRepository.cs [hunk 3]
```diff
@@ -32,4 +34,21 @@ public DocumentRepository(PortalDbContext dbContext)
...
+            .SelectMany(iamUser => iamUser.CompanyUser!.Documents.Where(docu => docu.DocumentTypeId == documentTypeId))
+            .Select(document =>
+                new UploadDocuments(
+                    document!.Id,
+                    document!.DocumentName))
+            .AsAsyncEnumerable();
+
+    /// <inheritdoc />
+    public async Task<Document?> GetDocumentByIdAsync(Guid documentId) =>
+        await this._dbContext.Documents.SingleOrDefaultAsync(x => x.Id == documentId);
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/DocumentRepository.cs:53]: no need to await this task here, just return it (removing 'async' from the signature)

File: src/registration/CatenaX.NetworkServices.Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 4]
```diff
@@ -85,6 +85,26 @@ public async Task<int> UploadDocumentAsync(Guid applicationId, IFormFile documen
...
+            {
+                throw new NotFoundException("No document with the given id was found.");
+            }
+
+            if (document.CompanyUserId is not null && document.CompanyUserId != userId)
+            {
+                throw new ForbiddenException("User doesn't have the relevant rights to request for the document");
+            }
+
+            return (document.DocumentName, Convert.FromBase64String(document.DocumentContent));
```
Comment (ntruchsess) [inline src/registration/CatenaX.NetworkServices.Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs:105]: document is byte[] not base64

File: src/registration/CatenaX.NetworkServices.Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 5]
```diff
@@ -85,6 +85,26 @@ public async Task<int> UploadDocumentAsync(Guid applicationId, IFormFile documen
             }
         }
 
+        public async Task<(string fileName, byte[] content)> GetDocumentAsync(Guid documentId, string iamUserId)
+        {
+            var userRepository = _portalRepositories.GetInstance<IUserRepository>();
+            var documentRepository = _portalRepositories.GetInstance<IDocumentRepository>();
+
+            var userId = await userRepository.GetCompanyUserIdForIamUserUntrackedAsync(iamUserId).ConfigureAwait(false);
+            var document = await documentRepository.GetDocumentByIdAsync(documentId).ConfigureAwait(false);
```
Comment (ntruchsess) [inline src/registration/CatenaX.NetworkServices.Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs:94]: as the document-content is rather huge it seems to be more efficient to first query the iamuserid that is assigned to this document to do the authorization-check first, then retrieve the content in separate query. Even if this is an additional call it saves a lot of resources if an attacker would try to DDos this service.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/DocumentsBusinessLogic.cs (hunk 2 above)
Comment (Phil91) [reply src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/DocumentsBusinessLogic.cs]: fixed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/DocumentRepository.cs (hunk 3 above)
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/DocumentRepository.cs]: fixed

File: src/registration/CatenaX.NetworkServices.Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs (hunk 4 above)
Comment (Phil91) [reply src/registration/CatenaX.NetworkServices.Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs]: fixed

File: src/registration/CatenaX.NetworkServices.Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs (hunk 5 above)
Comment (Phil91) [reply src/registration/CatenaX.NetworkServices.Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs:105]: good idea. i changed it

**Review by ntruchsess — APPROVED** [review]
@Phil91 I did some cleanup, please double-check. Feel free to merge if your are fine with my changes

**Comment by Phil91** [comment]
> @Phil91 I did some cleanup, please double-check. Feel free to merge if your are fine with my changes

Looks fine by me. Will merge it

**Pull request merged**